### PR TITLE
feat: Added custom webhook support for sending notification

### PR DESCRIPTION
**Pull request by adi6859**
# Description
Right now Devtron only supports sending notifications on SES, SMTP, Slack channel. Users might want to configure a custom webhook with configurable payload. New channel required to configure webhook.

Fixes [#AB3456](https://github.com/devtron-labs/devtron/issues/3456)

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
-all test are done locally 
-testing is also done by deploying image on vm

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: internal/sql/repository/WebhookNotificationRepository.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/devtron-labs/devtron/pkg/sql"
+	"github.com/go-pg/pg"
+)
+
+type WebhookNotificationRepository interface {
+	FindOne(id int) (*WebhookConfig, error)
+	UpdateWebhookConfig(webhookConfig *WebhookConfig) (*WebhookConfig, error)
+	SaveWebhookConfig(webhookConfig *WebhookConfig) (*WebhookConfig, error)
+	FindAll() ([]WebhookConfig, error)
+	FindByIdsIn(ids []int) ([]*WebhookConfig, error)
```
Comment (vikramdevtron) [inline internal/sql/repository/WebhookNotificationRepository.go:13]: remove functions which are not in use.

File: internal/sql/repository/WebhookNotificationRepository.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+func (impl *WebhookNotificationRepositoryImpl) FindOne(id int) (*WebhookConfig, error) {
+	details := &WebhookConfig{}
+	err := impl.dbConnection.Model(details).Where("id = ?", id).
+		Where("deleted = ?", false).Select()
+
+	return details, err
+}
+
+func (impl *WebhookNotificationRepositoryImpl) FindAll() ([]WebhookConfig, error) {
+	var webhookConfigs []WebhookConfig
```
Comment (vikramdevtron) [inline internal/sql/repository/WebhookNotificationRepository.go:60]: use pointers

File: internal/sql/repository/WebhookNotificationRepository.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+func (impl *WebhookNotificationRepositoryImpl) UpdateWebhookConfig(webhookConfig *WebhookConfig) (*WebhookConfig, error) {
+	return webhookConfig, impl.dbConnection.Update(webhookConfig)
+}
+
+func (impl *WebhookNotificationRepositoryImpl) SaveWebhookConfig(webhookConfig *WebhookConfig) (*WebhookConfig, error) {
+	return webhookConfig, impl.dbConnection.Insert(webhookConfig)
+}
+
+func (impl *WebhookNotificationRepositoryImpl) FindByName(value string) ([]WebhookConfig, error) {
+	var webhookConfigs []WebhookConfig
```
Comment (vikramdevtron) [inline internal/sql/repository/WebhookNotificationRepository.go:90]: pointer

File: pkg/notifier/SlackNotificationService.go [hunk 4]
```diff
@@ -250,8 +264,10 @@ func (impl *SlackNotificationServiceImpl) RecipientListingSuggestion(value strin
 								result := &NotificationRecipientListingResponse{
 									Recipient: v.(string),
 								}
-								if strings.Contains(v.(string), "https://") {
+								if strings.Contains(v.(string), "https://hooks.slack.com/") {
```
Comment (vikramdevtron) [inline pkg/notifier/SlackNotificationService.go:267]: put in const

File: pkg/notifier/WebhookNotificationService.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	"github.com/go-pg/pg"
+	"go.uber.org/zap"
+	"time"
+)
+
+const WEBHOOK_CONFIG_TYPE = "webhook"
+
+type WebhookAttribute string
+
+const (
```
Comment (vikramdevtron) [inline pkg/notifier/WebhookNotificationService.go:20]: add comments, over these fields will be configurable in future

File: pkg/notifier/WebhookNotificationService.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+type WebhookNotificationServiceImpl struct {
+	logger                         *zap.SugaredLogger
+	webhookRepository              repository.WebhookNotificationRepository
+	teamService                    team.TeamService
+	userRepository                 repository2.UserRepository
+	notificationSettingsRepository repository.NotificationSettingsRepository
+}
+type WebhookChannelConfig struct {
+	Channel           util2.Channel      `json:"channel" validate:"required"`
+	WebhookConfigDtos []WebhookConfigDto `json:"configs"`
```
Comment (vikramdevtron) [inline pkg/notifier/WebhookNotificationService.go:51]: pointer

File: pkg/notifier/WebhookNotificationService.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+	webhookConfig, err := impl.webhookRepository.FindOne(id)
+	if err != nil && !util.IsErrNoRows(err) {
+		impl.logger.Errorw("cannot find all webhoook config", "err", err)
+		return nil, err
+	}
+	webhoookConfigDto := impl.adaptWebhookConfig(*webhookConfig)
+	return &webhoookConfigDto, nil
+}
+
+func (impl *WebhookNotificationServiceImpl) GetWebhookAttributes() (map[string]WebhookAttribute, error) {
```
Comment (vikramdevtron) [inline pkg/notifier/WebhookNotificationService.go:114]: can we replace Attributes >> Variables

File: pkg/notifier/WebhookNotificationService.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+	repository2 "github.com/devtron-labs/devtron/pkg/user/repository"
+	util2 "github.com/devtron-labs/devtron/util/event"
+	"github.com/go-pg/pg"
+	"go.uber.org/zap"
+	"time"
+)
+
+const WEBHOOK_CONFIG_TYPE = "webhook"
+
+type WebhookAttribute string
```
Comment (vikramdevtron) [inline pkg/notifier/WebhookNotificationService.go:18]: can we replace Attributes >> Variables

File: api/router/NotificationRouter.go [hunk 9]
```diff
@@ -63,6 +63,13 @@ func (impl NotificationRouterImpl) InitNotificationRegRouter(configRouter *mux.R
 	configRouter.Path("/channel/smtp/{id}").
 		HandlerFunc(impl.notificationRestHandler.FindSMTPConfig).
 		Methods("GET")
+	configRouter.Path("/channel/webhook/{id}").
+		HandlerFunc(impl.notificationRestHandler.FindWebhookConfig).
+		Methods("GET")
+	configRouter.Path("/attribute").
```
Comment (vikramdevtron) [inline api/router/NotificationRouter.go:69]: can we replace attributes >> variables

File: api/restHandler/NotificationRestHandler.go [hunk 10]
```diff
@@ -45,19 +45,21 @@ import (
...
 type NotificationRestHandler interface {
 	SaveNotificationSettings(w http.ResponseWriter, r *http.Request)
 	UpdateNotificationSettings(w http.ResponseWriter, r *http.Request)
-
 	SaveNotificationChannelConfig(w http.ResponseWriter, r *http.Request)
 	FindSESConfig(w http.ResponseWriter, r *http.Request)
 	FindSlackConfig(w http.ResponseWriter, r *http.Request)
 	FindSMTPConfig(w http.ResponseWriter, r *http.Request)
+	FindWebhookConfig(w http.ResponseWriter, r *http.Request)
+	GetWebhookAttributes(w http.ResponseWriter, r *http.Request)
```
Comment (vikramdevtron) [inline api/restHandler/NotificationRestHandler.go:62]: Attributes >> Variables

File: pkg/notifier/SlackNotificationService.go (hunk 4 above)
Comment (adi6859) [reply pkg/notifier/SlackNotificationService.go:267]: Done

File: pkg/notifier/WebhookNotificationService.go (hunk 5 above)
Comment (adi6859) [reply pkg/notifier/WebhookNotificationService.go:20]: done

File: pkg/notifier/WebhookNotificationService.go (hunk 6 above)
Comment (adi6859) [reply pkg/notifier/WebhookNotificationService.go:51]: done

File: pkg/notifier/WebhookNotificationService.go (hunk 8 above)
Comment (adi6859) [reply pkg/notifier/WebhookNotificationService.go:18]: done

File: internal/sql/repository/WebhookNotificationRepository.go (hunk 3 above)
Comment (adi6859) [reply internal/sql/repository/WebhookNotificationRepository.go:90]: done

File: pkg/notifier/WebhookNotificationService.go (hunk 7 above)
Comment (adi6859) [reply pkg/notifier/WebhookNotificationService.go:114]: done

File: internal/sql/repository/WebhookNotificationRepository.go (hunk 2 above)
Comment (adi6859) [reply internal/sql/repository/WebhookNotificationRepository.go:60]: done

File: internal/sql/repository/WebhookNotificationRepository.go (hunk 1 above)
Comment (adi6859) [reply internal/sql/repository/WebhookNotificationRepository.go:13]: done

File: internal/sql/repository/WebhookNotificationRepository.go (hunk 2 above)
Comment (adi6859) [reply internal/sql/repository/WebhookNotificationRepository.go:60]: done

**Pull request closed without merging**